llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -551,102 +686,98 @@ TableFiles processOneModule(std::unique_ptr<Module> M, bool IsEsimd,
-                                       : module_split::SPLIT_NONE;
-  std::unique_ptr<module_split::ModuleSplitterBase> MSplit =
-      module_split::getSplitterByMode(std::move(M), Mode, IROutputOnly,
-                                      EmitOnlyKernelsAsEntryPoints,
-                                      DeviceGlobals);
-
-  StringRef FileSuffix = IsEsimd ? "esimd_" : "";
-
-  for (size_t I = 0; I < MSplit->totalSplits(); ++I) {
-    module_split::ModuleDesc ResMDesc = MSplit->nextSplit();
-    Module &ResM = ResMDesc.getModule();
-
-    bool SpecConstsMet = processSpecConstants(ResM);
-    bool CompileTimePropertiesMet = processCompileTimeProperties(ResM);
-
-    if (IROutputOnly) {
-      // the result is the transformed input LLVM IR file rather than a file
-      // table
-      saveModuleIR(ResM, OutputFilename);
-      return TblFiles;
-    }
-
-    {
-      // Reuse input module with only regular SYCL kernels if there were
-      // no spec constants and no splitting.
-      // We cannot reuse input module for ESIMD code since it was transformed.
-      std::string ResModuleFile{};
-      bool CanReuseInputModule = !SyclAndEsimdCode && !IsEsimd &&
-                                 !IsLLVMUsedRemoved && !SpecConstsMet &&
-                                 !CompileTimePropertiesMet &&
-                                 (MSplit->totalSplits() == 1);
-      if (CanReuseInputModule)
-        ResModuleFile = InputFilename;
-      else {
-        ResModuleFile =
-            makeResultFileName((OutputAssembly) ? ".ll" : ".bc", I, FileSuffix);
-        saveModuleIR(ResM, ResModuleFile);
+  Modified |= removeSYCLKernelsConstRefArray(*M.get());
+
+  // Do invoke_simd processing before splitting because this:
+  // - saves processing time (the pass is run once, even though on larger IR)
+  // - doing it before SYCL/ESIMD splitting is required for correctness
+  Modified |= lowerInvokeSimd(*M);
+
+  DUMP_ENTRY_POINTS(*M, EmitOnlyKernelsAsEntryPoints, "Input");
+
+  // --ir-output-only assumes single module output thus no code splitting.
+  // Violation of this invariant is user error and must've been reported.
+  // However, if split mode is "auto", then entry point filtering is still
+  // performed.
+  assert(!IROutputOnly || (SplitMode == module_split::SPLIT_NONE) ||
+         (SplitMode == module_split::SPLIT_AUTO) &&
+             "invalid split mode for IR-only output");
+
+  // Top-level per-kernel/per-source splitter. SYCL/ESIMD splitting is applied
+  // to modules resulting from the top-level splitting.
+  std::unique_ptr<module_split::ModuleSplitterBase> Splitter =
+      module_split::getSplitterByMode(std::move(M), SplitMode, IROutputOnly,
+                                      EmitOnlyKernelsAsEntryPoints);
+  Modified |= Splitter->totalSplits() > 1;
+
+  if (DeviceGlobals)
+    Splitter->verifyNoCrossModuleDeviceGlobalUsage();
+
+  // Proceed with top-level splitting.
+  while (Splitter->hasMoreSplits()) {
```

Comment:
`processInputModule` now has two nested splitters - SYCL/ESIMD is nested within scoped one.